### PR TITLE
allow the /version endpoint to pass through

### PR DESCRIPTION
**Pull request by deads2k**
Makes `/version` a normal endpoint which doesn't use gorestful to make it easier to predictably aggregate and pass through to servers which have `/version/foo` endpoints.

@kubernetes/sig-api-machinery-misc

### CODE REVIEW

File: api/openapi-spec/swagger.json [hunk 1]
```diff
@@ -46070,35 +46070,6 @@
       "required": true
      }
     ]
-   },
-   "/version/": {
```
Comment (mbohlool) [inline api/openapi-spec/swagger.json]: Removing an endpoint from OpenAPI spec is a backward incompatible change. Is it possible that an OpenAPI client rely on this?

**Review by mbohlool — CHANGES_REQUESTED** [review]

File: api/openapi-spec/swagger.json (hunk 1 above)
Comment (deads2k) [inline api/openapi-spec/swagger.json]: > Removing an endpoint from OpenAPI spec is a backward incompatible change. Is it possible that an OpenAPI client rely on this?

Eh, maybe?  It's not our general intent.  How is generation from something other than gorestful coming?  That library is really problematic.

Comment (mbohlool) [inline api/openapi-spec/swagger.json]: It is not only OpenAPI. It's old swagger 1.2 too. We need to support it 1.8. After 1.9 release we can remove gorestful along with old swagger 1.2 and have our inhouse http routers or some other libraries. Until then, I don't think it is a good idea to move endpoints outside of gorestful. Can you elaborate on the problem this causes for aggregation?

Comment (deads2k) [inline api/openapi-spec/swagger.json]: > Can you elaborate on the problem this causes for aggregation?

Subpaths under `/version` are hidden when chaining servers together.  It means that you can't have `/version/apiextensions-apiserver` or `/version/openshift`.

Comment (mbohlool) [inline api/openapi-spec/swagger.json]: That means those subpaths are not registered with a webservice and does not have Swagger/OpenAPI spec? From the code (both old and new) I can only see registering `/version`. Where do we register `/version/apiextensions-apiserver` etc?

Comment (deads2k) [inline api/openapi-spec/swagger.json]: > That means those subpaths are not registered with a webservice and does not have Swagger/OpenAPI spec? From the code (both old and new) I can only see registering /version. Where do we register /version/apiextensions-apiserver etc?

`/version/apiextensions-apiserver` is theoretical, but probably necessary at somepoint.  `/version/openshift` exists in the openshift apiserver which is delegated in the chain.

Comment (mbohlool) [inline api/openapi-spec/swagger.json]: Sorry if I still don't see the problem. `/version/openshift` and `/version/apiextensions-apiserver` can be registered the same way as `/version` in both cases. How gorestful hides those endpoints and the new implementation doesn't?

Comment (deads2k) [inline api/openapi-spec/swagger.json]: > Sorry if I still don't see the problem. /version/openshift and /version/apiextensions-apiserver can be registered the same way as /version in both cases. How gorestful hides those endpoints and the new implementation doesn't?

gorestful forces a 404 when a request is made the webservice, but not matching a route. Our "normal" mux doesn't do that.

Comment (mbohlool) [inline api/openapi-spec/swagger.json]: I see. I am going back to my previous question then. That means those subpaths are not registered with a webservice and does not have Swagger/OpenAPI spec? Is it not a better idea to just register them with a webservice?

Comment (mbohlool) [inline api/openapi-spec/swagger.json]: or if we don't want to limit the tail of the url, we can always use `*` like this example: https://github.com/emicklei/go-restful/blob/master/examples/restful-path-tail.go.

**Pull request closed without merging**